### PR TITLE
Add podman-os build job

### PR DESCRIPTION
**Pull request by ravanelli**
- Integrate essential Podman build jobs into our CI
infrastructure to facilitate unblocking tasks for
the Podman team. The objective is to ensure
flexibility by avoiding strong dependencies on our
internal components such as cosa and libraries,
enabling seamless transition for the Podman team in the future.

### CODE REVIEW

**Comment by ravanelli** [comment]
I still need the podam repo names so I can replace from mine

**Review by jlebon — COMMENTED** [review]
First, thank you for working on this!

I haven't done a detailed review yet, but one thing that jumped out is the use of two separate jobs. Notably, we're building the x86_64 artifacts first, and then the aarch64 artifacts in a separate job. I think ideally it would all be in a single job instead. This keeps it self-contained, but also e.g. things like manifest list building become easier if it's all in a single job.

See e.g. how the `bump-lockfile` or [`test-override`](https://github.com/coreos/coreos-ci/blob/main/jobs/test-override.Jenkinsfile) jobs are structured.

OK, here's a rough outline of what I mean:

```groovy
cosaPod(...) {
	// build images for every arch, push to some staging repo
	stage("Build Layered Image") {
		parallel arches.collectEntries{arch -> [arch, {
			cosa remote-build-container ...
		}]}
	}
	// now we build disk images
	stage("Initialize Remotes") {
		parallel arches.collectEntries{arch -> [arch, {
			if (arch != x86_64) {
				cosa remote-session create
				// store remote ID in e.g. archinfo map
			}
		}]}
	}
	stage("Build Disk Images") {
		parallel arches.collectEntries{arch -> [arch, {
			withOptionalExistingCosaRemoteSession(arch: arch, session: archinfo[arch]['session'] {
				cosa supermin-run ...
			}
		}]}
	}
	// convert disk images to container images
	// push all the container images to some staging repo
	// build manifest lists
	// push out manifest lists to final repo
	// clean up staging repo
}
```

This also avoids copying any container image between builders.

WDYT?

**Comment by ravanelli** [comment]
In fact, this was my first approach, two arches in one  job, but thinking about the case of an arch failing, this would limit the rebuild part, so I decided to split it and follow the way we do with our multiarch ones

**Comment by jlebon** [comment]
> In fact, this was my first approach, two arches in one job, but thinking more about the case of an arch failing, this would limit the rebuild part

This is a good point. I think we should be able to hit the cache at least for the building of the layered images stage. You're right we'd probably lose the work on the disk images though. That said, since we're not e.g. running the full testsuite, I think in practice this job will be quite stable (just like e.g. `build-cosa` is pretty stable today and mostly flakes on infra issues).

> and both jobs can run almost in parallel

In the suggested approach, we'd be parallelizing across arches right from the start, so it *should* be faster too. :) (Though I think the speed should be quite reasonable either approach overall.)

> What do you mean by: This also avoids copying any container image between builders?

IIUC, we're doing `cosa remote-build-container ... --write-to-file ...` which will transfer the OCI archive from the builder to Jenkins, and then later on `cosa remote-session sync` to sync it back to the remote, right? I think ideally we should push directly to the staging repo from the remote builders (both for the OS container and the disk images) to avoid incurring that cost.

>  I'm fine with this way too!

:+1: 

I think you have all the key pieces here, so it'd just be about re-arranging them a bit. :) If you want to defer this and try to get this approach in first, I'm OK with that too. If you want to get together to try to hammer this out, we can do that too.

**Comment by lsm5** [comment]
> I still need the podam repo names so I can replace from mine

@ravanelli thanks, you can use `quay.io/podman/staging` and `quay.io/podman/machine-images` for now. Also, the secret can be replaced with `podman-push-registry-secret`

**Comment by ravanelli** [comment]
We still need to define the manifest final merge, if it is not possible I will need the artifacts back to the host to then create the full manifest with all arches on it.

Here is how the staging looks like:
https://quay.io/repository/ravanelli/coreos-assembler-staging

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:121]: I think we can probably bring this down by a lot. Especially since we're using zstd which I don't think is as memory hungry.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:129]: This too.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:2]: This is only used in the `lock()`; seems OK to drop this variable and hardcode "build-podman-fcos" in the `lock()`?

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:171]: This is important; probably worth hoisting to the top definitions for more visibility?

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:226]: This part could be a separate step from remote session initialization.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:232]: We shouldn't need this conditional. `withOptionalExistingCosaRemoteSession()` already knows to directly execute the passed body directly if it's x86_64 (well really, if no `session` is defined). So this would just be:

```
parallel ... {
  pipeutils.withOptional...() {
    generate_disk_images(...)
  }
}
```

And similarly for the other stages.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:248]: Aside: keep an eye on https://github.com/coreos/coreos-assembler/pull/3652; we'll have to update this once that PR merges.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:264]: The canonical way to run arbitrary commands is `cosa shell --`. That'll work both in the remote and non-remote case (related to the previous comment about `withOptionalExistingCosaRemoteSession()`).

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:190]: Ideally we would limit the credential scope to the stages that need it.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:374]: This needs updating.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:288]: Similarly, I think we should be able to get rid of this.

File: jobs/build-podman.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman.Jenkinsfile:283]: So actually, we do have a nice wrapper for this: [`utils.syncCredentialsIfInRemoteSession()`](https://github.com/coreos/coreos-ci-lib/blob/cf8e6a14b2c8b6d83940aa3f20ab6847345fd379/vars/utils.groovy#L65). You can see sample usage in other code in this repo.

**Review by jlebon — COMMENTED** [review]
This is looking really great! :+1:

**Comment by lsm5** [comment]
> We still need to define the manifest final merge, if it is not possible I will need to sync the artifacts back to the host to then create the full manifest with all arches on it.
> 
> Here is how the staging looks like: https://quay.io/repository/ravanelli/coreos-assembler-staging?tab=tags
> 
> @lsm5 @baude, also please validate the release manifest part, is that what you need? We can keep pushing to staging until we decided to move it to the official repo.

@baude PTAL

File: jobs/build-podman.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman.Jenkinsfile:25]: ouch.. this isn't the most ergonomic because of the way runvm-osbuild essentially takes one platform/filename at a time. I'll maybe see if there is way to make it better so we can just create all the platforms with one single run.

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman-os.Jenkinsfile:297]: in the future we should look into trying to do this outside of supermin.

right now in FCOS we don't use supermin when combining into a manifest list and pushing to the registry:

https://github.com/coreos/fedora-coreos-pipeline/blob/6eebb70044dc67a349e826006ebcbf922e8b5b24/jobs/release.Jenkinsfile#L267-L271

which just uses https://github.com/coreos/coreos-assembler/blob/main/src/cmd-push-container-manifest


we also don't use it when building cosa:

https://github.com/coreos/fedora-coreos-pipeline/blob/6eebb70044dc67a349e826006ebcbf922e8b5b24/jobs/build-cosa.Jenkinsfile#L131-L154


Maybe the `export STORAGE_DRIVER=vfs` is the key.

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman-os.Jenkinsfile:316]: can you help me understand what is going on here behind the scenes?

What is the difference between these two?:
- staging to main repo `5.0-${shortcommit} -> 5.0`
- for each arch
    - staging to main repo `${arch}-${shortcommit} -> 5.0-${arch}`

**Comment by dustymabe** [comment]
also, why did you drop the `stage('Delete Intermediate Tags')`? was it not working? or do we ant to keep them around for some reason?

File: jobs/build-podman-os.Jenkinsfile
Comment (ravanelli) [reply jobs/build-podman-os.Jenkinsfile:297]: The `build-cosa` doesn't use kvm. Since we need supermin to run osbuild we need to enable it. Seems it is causing issues with podman. For `release` the  `push-container-manifest` uses skopeo

**Comment by ravanelli** [comment]
> also, why did you drop the `stage('Delete Intermediate Tags')`? was it not working? or do we ant to keep them around for some reason?
Kind of, if we clean up staging we lose the ability in avoiding to build the container image if it is already there. As we can also add a check  further for the artifacts builds as well. I was thinking in maybe adding an expiration tag for it.

**Comment by dustymabe** [comment]
> Kind of, if we clean up staging we lose the ability in avoiding to build the container image if it is already there. As we can also add a check further for the artifacts builds as well. I was thinking in maybe adding an expiration tag for it.

hmm. if we get to the point where we delete the intermediate tags then we got all the way to success right? so we wouldn't ever need them again?

We do this today in the build-cosa job.. if the job fails part way through then it won't clean up the intermediate tags and if some arches were successful then it will just re-use those. If it gets all the way to success then it cleans them up.

**Comment by ravanelli** [comment]
>> hmm. if we get to the point where we delete the intermediate tags then we got all the way to success right? so we wouldn't ever need them again?
Yes!

>> We do this today in the build-cosa job.. if the job fails part way through then it won't clean up the intermediate tags and if some arches were successful then it will just re-use those. If it gets all the way to success then it cleans them up.
Yeah, that's make sense. I'm fine adding it back (will do)

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [reply jobs/build-podman-os.Jenkinsfile]: yes. I think the question I'm asking is: why?

why do we need that? are the users going to use them directly? is it some artifact of the way we are pushing that they need to be there?

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [reply jobs/build-podman-os.Jenkinsfile:297]: > Since we need supermin to run osbuild we need to enable it.

right. we need supermin for when we run `osbuild`, but we're not running `osbuild` here, we're running `buildah`. 



> For `release` the `push-container-manifest` uses skopeo

right. `podman`, `buildah`, `skopeo` all use the same libraries underneath. I suspect we can figure out how to run `buildah manifest` commands here without `supermin` too, but we don't need to block on that.

File: jobs/build-podman.Jenkinsfile
Comment (dustymabe) [reply jobs/build-podman.Jenkinsfile]: this is now 24h

File: jobs/build-podman.Jenkinsfile
Comment (dustymabe) [reply jobs/build-podman.Jenkinsfile]: looks like this was updated

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman-os.Jenkinsfile:36]: I feel like this value needs to be updated?

File: jobs/build-podman-os.Jenkinsfile
Comment (ravanelli) [reply jobs/build-podman-os.Jenkinsfile]: I think it is a question for @baude

Comment (baude) [reply jobs/build-podman-os.Jenkinsfile]: @dustymabe i dont get what you are asking me, ill reach out

Comment (dustymabe) [reply jobs/build-podman-os.Jenkinsfile]: ok talked with brent,

basically we need the `5.0` tag to include the oci artifacts too (i.e. the containers and not just the disk images) so we'll need to add a few more steps to the `buildah manifest` command run in the supermin VM. 

we don't actually need the `5.0-aarch64` and `5.0-x86_64` in the registry so we can just not push those (i.e. we can delete the second `skopeo copy`).

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman-os.Jenkinsfile:96]: From talking with brent I think this is where he is pushing too:

```suggestion
             defaultValue: "quay.io/podman/machine-os",
```

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman-os.Jenkinsfile:36]: if we do https://github.com/coreos/fedora-coreos-pipeline/pull/968/files#r1531138529 then this will need updating too

**Review by lsm5 — COMMENTED** [review]
lgtm as far as i can tell. The repo paths look ok too, but maybe `quay.io/podman/staging` should be replaced with something more specific like: `quay.io/podman/machine-os-staging` ? @baude if you agree, I can set that up and update here.

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [inline jobs/build-podman-os.Jenkinsfile:298]: is the `docker://` required here? 
```suggestion
                    ${manifest}"
```

Looking at the COSA code that pushes I don't think we have `docker://` on there.

**Review by dustymabe — APPROVED** [review]
LGTM

File: jobs/build-podman-os.Jenkinsfile
Comment (ravanelli) [reply jobs/build-podman-os.Jenkinsfile:298]: If I recall buildah complains about the transport if we don't add it.

Comment (dustymabe) [reply jobs/build-podman-os.Jenkinsfile:298]: ok

**Review by dustymabe — APPROVED** [review]

File: jobs/build-podman-os.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman-os.Jenkinsfile:159]: Minor/optional: comment looks out of date -- delete?

File: jobs/build-podman-os.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman-os.Jenkinsfile:229]: Wait, why are we doing another `cosa init` here? The one above should work fine across all arches.

File: jobs/build-podman-os.Jenkinsfile
Comment (jlebon) [inline jobs/build-podman-os.Jenkinsfile:290]: Minor/optional:

```suggestion
                //  `cosa supermin-run` only mounts the current directory, so we
                // need to temporarily put the secret here. We'll delete it right after.
```

**Review by jlebon — COMMENTED** [review]
@ravanelli Nice work! This looks much nicer now.

Only two comments I think we should address. The rest is a string of minor comments that you're free to ignore/address in a follow up.

File: jobs/build-podman-os.Jenkinsfile
Comment (dustymabe) [reply jobs/build-podman-os.Jenkinsfile:164]: TBH I'm pretty sure we're not actually using this (or rather needing this) here becuase we are using runvm-osbuild which sets `outdir=cache/osbuild/out` so it's the same FS and thus no copying over virtiofs.

File: jobs/build-podman-os.Jenkinsfile
Comment (ravanelli) [reply jobs/build-podman-os.Jenkinsfile:274]: Yes, we are still defining from where it will come from, that's why I keept  it hard code for now

**Comment by dustymabe** [comment]
I think my `OSBUILD_EXPORT_FORCE_NO_PRESERVE_OWNER` comment is still outstanding: https://github.com/coreos/fedora-coreos-pipeline/pull/968#discussion_r1537680287

We don't need to remove it, but it is confusing to have it when we don't need it.

**Review by dustymabe — APPROVED** [review]

File: jobs/build-podman-os.Jenkinsfile
Comment (ravanelli) [reply jobs/build-podman-os.Jenkinsfile]: I missed it before, fixed

**Review by dustymabe — APPROVED** [review]
LGTM

**Comment by lsm5** [comment]
@ravanelli @jlebon @dustymabe thanks a lot for this!!

**Pull request merged**
